### ISSUE TITLE
Generate meaningful SegmentIDs instead of pure random

### ISSUE THREAD

**Issue by ppodolsky** [quickwit-oss/tantivy#971]
**Is your feature request related to a problem? Please describe.**
Related to #969 
I would like to suggest a cheap feature which will help debugging in the future.
Now SegmentIDs are generated randomly but it is a waste of 16 bytes which could be used to embed debugging info otherwise.

**Describe the solution you'd like**
Generate SegmentID containing the following info: 
- timestamp of segment creation
- segment origin (merging or writing new data)
- hash of hostname, probably useful for those who will implement sharding/replication paired with Tantivy.

Additionally here we should ensure that there is left enough randomness to avoid any possibility of collisions on the one hand, and that names are not too long to avoid metadata bloating (? not sure if it is actual, the number of segments is supposed to be relatively low by design afaik) on the other hand.

**Comment by scampi** [comment]
I would be interested working on this feature, but I am uncertain how to move forward.

> hash of hostname, probably useful for those who will implement sharding/replication paired with Tantivy.

This sounds to me like a feature outside of the scope of Tantivy and better suited to projects that implement a distributed search engine over it.

> segment origin (merging or writing new data)

From what I understand of segment merging, I would have the following layout most of the time:

```
segment-1-merge
segment-2-new
segment-3-new
...
segment-20-new
```

It feels like this information could be deduced from the created timestamp of a file: the oldest file is probably one that got merged, which could be retrieved by executing `ls -l`. However, with this alone I can't be sure if the oldest file results from a merge.

I had a look at Lucene for some inspiration: from what I can tell, a segment name is created by [IndexWriter#newSegmentName](https://github.com/apache/lucene/blob/d5d6dc079395c47cd6d12dcce3bcfdd2c7d9dc63/lucene/core/src/java/org/apache/lucene/index/IndexWriter.java#L2044) which increases a segment counter.

It seems that a counter would resolve the previous shortcoming: if the counter is 1, then for sure the oldest file is a newly created segment.

What about a segment ID that is just a generation counter ? The nice aspect however of using `uuid` at the moment is that a segment can be created by any thread, no sync needed.

**Comment by fulmicoton** [comment]
> This sounds to me like a feature outside of the scope of Tantivy and better suited to projects that implement a distributed search engine over it.

Well we could consider passing an optional string to the IndexWriter, but I agree #999 would be a much better fit.
Let's try and properly fleshout an exhaustive list of stuff we could want to see in the segment name and in segment metas,
and then arbitrage this.


@ppodolsky @scampi @guilload  @fmassot 

- meaningful lexicographical sort (sort by commit timestamp)
- a label given to the indexwriter (hostname)... But then would happens when merging two segments with different labels? Do we really want to enforce siloing these labels?
- segment generation
- previous generation ids
- commit id (due to multithreading... a commit gives birth to more than one segment)
- obviously we want it to be universally unique
- a not-so unique human friendly label to help human memorize and communicate labels when investigating a bug

What else ?

**Comment by ppodolsky** [comment]
Is it worth to include Tantivy version? Not footer version but Tantivy one's.

> But then would happens when merging two segments with different labels? Do we really want to enforce siloing these labels?

What do you mean by siloing? Keeping isolated sets of labels of all ancestors of the segment?

Options: 
- Erase labels
- Overwrite labels by what is set for the index_writer doing current merging
- Allow users to define a policy of label merging

**Comment by ppodolsky** [comment]
> This sounds to me like a feature outside of the scope of Tantivy and better suited to projects that implement a distributed search engine over it.

A naming policy of segments is fully managed by Tantivy and there is no way to highjack for any other project into it. It should be considered from the point of making debugging easier, details could be seen in #969.

So I'd like to suggest to make a decision. 

@fulmicoton told that segment attributes are better approach so what should be left for embedding into name of segments?
Probably it is worth to shorten the list down to the most important ones:

- Unixtime of generation
- Natural or merging origin

**Comment by scampi** [comment]
> > This sounds to me like a feature outside of the scope of Tantivy and better suited to projects that implement a distributed search engine over it.
> 
> A naming policy of segments is fully managed by Tantivy and there is no way to highjack for any other project into it. It should be considered from the point of making debugging easier, details could be seen in #969.

@ppodolsky I agree, I wrote this comment regarding the hashing of the hostname, not about the naming of segments.

**Comment by Divyesh-k** [comment]
Rather than encoding a hostname/label or merge-origin into the id (the parts of this thread that stalled on the "what happens to labels on merge?" siloing question), I took the small, uncontroversial slice: switch segment id generation from random UUIDv4 to time-ordered UUIDv7 (RFC 9562).

Why this is better than the original random v4:

Embedded creation timestamp — UUIDv7 puts a 48-bit millisecond creation time in the most significant bits. This is exactly the "unixtime of generation" item from the discussion, with no format change and no extra bytes.
Chronologically sortable — SegmentId's Ord is a byte comparison, so ids now sort by creation time. That was @fulmicoton's very first wishlist bullet ("meaningful lexicographical sort by commit timestamp").
Self-describing for debugging — added SegmentId::creation_time() -> Option<SystemTime> to recover the timestamp, so you no longer need to rely on filesystem ls -l guessing to tell when/what order segments were produced.
Why it doesn't reintroduce the old blockers:

Still universally unique and still lock-free from any indexing thread — the exact properties UUIDv4 was originally chosen for. This sidesteps the generation-counter idea, which would have needed cross-thread synchronization.
The counter-vs-uuid tension that stalled this in 2021 is simply obsoleted by UUIDv7 existing now (it was standardized after this issue was filed).
Backward compatible: on-disk format is unchanged (32-char hex uuid; v4/v7 differ only in the version nibble). Old indices keep working; creation_time() returns None for their v4 ids.

I deliberately left the hostname/label and richer origin metadata out of the id — per @fulmicoton's point, that's a better fit for segment attributes on SegmentMeta (#999) and avoids the unresolved label-merge policy. Happy to follow up with that separately if there's interest.

**Comment by Divyesh-k** [comment]
If there's a preference not to change the id generation itself, here's an alternative approach that reaches the same debugging goal from the other direction discussed here — putting the metadata in SegmentMeta instead of in the id (@fulmicoton's suggested fit, #999):

Add an optional attributes blob to SegmentMeta. Concretely, an extra field on InnerSegmentMeta:
struct InnerSegmentMeta {
    segment_id: SegmentId,
    max_doc: u32,
    pub deletes: Option<DeleteMeta>,
    // ...
    /// Optional writer-supplied debug attributes.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    attributes: Option<serde_json::Value>,
}
Set from the IndexWriter, which already knows its own label/hostname and whether it's producing a fresh segment vs a merge result — so the "natural vs merging origin" item is answered directly, no ls -l timestamp guessing.
Merge policy: the simplest defensible rule that avoids the siloing debate — the child segment gets the attributes of the writer doing the merge (the "overwrite" option from earlier in this thread); ancestors' labels aren't combined. A pluggable policy can come later.
skip_serializing_if = "Option::is_none" keeps meta.json byte-identical for anyone who doesn't set attributes → fully backward compatible.
Trade-offs vs the UUIDv7 PR:

➕ More flexible — can carry hostname, tantivy version, origin, arbitrary labels, not just a timestamp.
➖ Doesn't make ids sortable, and the info lives in meta.json rather than in the filename itself.
The two are complementary, not mutually exclusive — UUIDv7 gives free timestamp + sort in the id; attributes give the richer, opt-in metadata. Ideal end state is probably both.
I'm happy to implement whichever direction you'd prefer (or both). Just let me know.

**Comment by Divyesh-k** [comment]
@PSeitz @fulmicoton Can you please check and let mw know ?
